Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -17,6 +17,108 @@ import NIO
 let crlf: StaticString = "\r\n"
 let headerSeparator: StaticString = ": "
 
+/// An `IteratorProtocol` that can iterate through comma separated list of values for a certain
+/// header
+///
+///  **Example:**
+///
+/// Suppose you have this headers:
+///
+///      Connection: keep-alive, x-server
+///      Content-Type: text/html
+///      Connection: other
+///
+/// You can iterate using this struct on those headers, for values of `Connection`, to get
+/// `keep-alive`, then `x-server`, then `other`
+///
+public struct HTTPListHeaderIterator<T: Collection>: IteratorProtocol where T.Element == UInt8 {
+    
+    public typealias Element = ByteBufferView
+    
+    private var currentHeaderIndex: Int = -1
+    private var byteBuffer: ByteBuffer
+    private var singleValueViewIterator: Array<ByteBufferView>.Iterator?
+    private let headerName: T
+    private let headers: [HTTPHeader]
+    
+    fileprivate let comma = UInt8(",".utf8CString[0])
+    
+    /// Returns next index in headers
+    ///
+    /// - Parameter current: The index to begin iteration at
+    /// - Returns: The next index of the header in header array, or `nil` if not found
+    internal func headerIndex(after current: Int) -> Int? {
+        for (idx, currentHeader) in headers.enumerated().dropFirst(current + 1) {
+            if byteBuffer.viewBytes(at: currentHeader.name.start,
+                                    length: currentHeader.name.length)
+                .compareReadableBytes(to: headerName) {
+                return idx
+            }
+        }
+        return nil
+    }
+
+    mutating public func next() -> ByteBufferView? {
+        if let next = self.singleValueViewIterator?.next() {
+            return next.trimSpaces
+        } else {
+            // End of this buffer. Let's try to grab the next one.
+            guard let index = self.headerIndex(after: currentHeaderIndex) else {
+                // No more buffers left.
+                return nil
+            }
+            self.currentHeaderIndex = index
+            self.singleValueViewIterator = byteBuffer
+                .viewBytes(at: headers[currentHeaderIndex].value.start,
+                           length: headers[currentHeaderIndex].value.length)
+                .split(separator: comma)
+                .makeIterator()
+            return self.next()
+        }
+        
+    }
+    
+    init(headerName: T,
+         headers: HTTPHeaders) {
+        
+        self.headers = headers.headers
+        self.headerName = headerName
+        
+        self.byteBuffer = headers.buffer
```

Comment:
I don't think it's worth unpacking the `HTTPHeaders` structure here. We're never mutating any of the sub-fields like `headers` or `buffer`, so we can safely extract them from the structure as needed.